### PR TITLE
MacOS fix for failing tests in the test_zmq.py 

### PR DESCRIPTION
**Pull request by skarakuzu**
8 tests in the test_zmq.py has been failing on MacOs system and it seems like there is a problem in pickling done in the `multiprocessing` module. The problems seems to be resolved by importing `multiprocess` which seems to be using dill  instead as suggested in [this example](https://stackoverflow.com/questions/72766345/attributeerror-cant-pickle-local-object-in-multiprocessing)
We run all tests both on Linux and MacOS after this change and problem seems to be solved.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -26,6 +26,7 @@ dependencies = [
     "typing-extensions>=4.0.0;python_version<'3.11'",
     "dataclasses;python_version<'3.7'",
     "zict<3.0.0",
+    "multiprocess",
```
Comment (tacaswell) [inline pyproject.toml:29]:
```suggestion
```

I think we only need this in the dev section?

**Review by danielballan — CHANGES_REQUESTED** [review]
I misread. The `multiprocess` package needs to be added to the `dev` section of the dependencies.

**Comment by danielballan** [comment]
Test failure on one build looks like a fluke, and clearly unrelated to the changes in this PR:

```
=========================== short test summary info ============================
FAILED src/bluesky/tests/test_new_examples.py::test_descriptor_layout_from_monitor[True] - AssertionError: assert {'rand': 0.6287866689704653} == {'rand': 0.26557044305186606}
  
  Differing items:
  {'rand': 0.6287866689704653} != {'rand': 0.26557044305186606}
  
  Full diff:
    {
  -     'rand': 0.26557044305186606,
  +     'rand': 0.6287866689704653,
    }
= 1 failed, 1612 passed, 15 skipped, 18 xfailed, 16 xpassed in 1169.53s (0:19:29) =
ERROR: InvocationError for command /opt/hostedtoolcache/Python/3.10.14/x64/bin/pytest --cov=bluesky --cov-report term --cov-report xml:cov.xml (exited with code 1)
___________________________________ summary ____________________________________
ERROR:   tests: commands failed
```

**Comment by skarakuzu** [comment]
> Test failure on one build looks like a fluke, and clearly unrelated to the changes in this PR:
> 
> ```
> =========================== short test summary info ============================
> FAILED src/bluesky/tests/test_new_examples.py::test_descriptor_layout_from_monitor[True] - AssertionError: assert {'rand': 0.6287866689704653} == {'rand': 0.26557044305186606}
>   
>   Differing items:
>   {'rand': 0.6287866689704653} != {'rand': 0.26557044305186606}
>   
>   Full diff:
>     {
>   -     'rand': 0.26557044305186606,
>   +     'rand': 0.6287866689704653,
>     }
> = 1 failed, 1612 passed, 15 skipped, 18 xfailed, 16 xpassed in 1169.53s (0:19:29) =
> ERROR: InvocationError for command /opt/hostedtoolcache/Python/3.10.14/x64/bin/pytest --cov=bluesky --cov-report term --cov-report xml:cov.xml (exited with code 1)
> ___________________________________ summary ____________________________________
> ERROR:   tests: commands failed
> ```

This error never happened on my computer or in the previous build. What would be the reason for the random number change? I do not think that I touched any fixed seed for the random number generator in this PR.

**Comment by prjemian** [comment]
I, too, have observed such random failures in other GHA workflows.  Re-run the failed tests (from the Actions tab) and check if failure repeats.

**Review by mrakitin — COMMENTED** [review]
Looks good, thanks for your contribution, @skarakuzu!

Would it be too bad to include `macos-13` and `macos-14` workers in the test matrix?
@danielballan, @coretl, @tacaswell, @callumforrester?

**Comment by tacaswell** [comment]
made a new issue for extending the CI matrix

**Comment by mrakitin** [comment]
Thanks, @tacaswell!
Testing it locally now on both macos amd64 and arm64 architectures.

**Comment by danielballan** [comment]
As the CI is getting backed up today, it is a reminder that there is eventually a cost to stacking these up. And Bluesky in particular in mainly run on Linux, mostly only _developed_ on Mac. Perhaps a single Mac build on the latest supported Python version would be a pragmatic compromise. (This is what we do for Windows builds with Tiled.)

**Comment by danielballan** [comment]
Ah, my comment crossed with Tom's. Will comment on new issue.

**Comment by mrakitin** [comment]
Agreed, a single run on each platform will help. No need for the entire matrix to run there.

Just wanted to clarify why we may need two separate macOS runners: the `macos-13` is designed to run OSX x86_64 tests, and `macos-14` is for the new Apple Silicon arm64 arch.

**Comment by mrakitin** [comment]
I don't understand the codecov failure. https://app.codecov.io/gh/bluesky/bluesky/pull/1725 seems happy/green. Do we care about the reduction of the coverage?

**Review by mrakitin — APPROVED** [review]
Thanks for the fix, @skarakuzu! The tests now pass locally on MacOS arm64.

**Comment by mrakitin** [comment]
> The tests seems to take a lot of time to complete. Is there any parallelization schema applied currently? If not, is it possible for the code to handle tests running asynchronously?

I think it's because of the parallel activities - multiple PRs are submitted in parallel, and we only have 10 concurrent jobs to run at a time. I remember @coretl mentioned the `tox` executor can run tests in parallel, but I am not sure it's configured to do so now.

**Comment by tacaswell** [comment]
In term of parallizing the actual test run, you can do `pytest -n 20` or what ever, which does run most of the tests in parallel between a bunch of workers.  However, the zmq tests (which are the slowest due to long waits) all use the same ports so are not safe to run in parallel.

There are also limits to how much parallelism buys you an CI as we only have 2 cores, running 10 workers does not buy you anything (and may put memory pressure and get you oom'd).

**Pull request merged**